versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -0,0 +1,936 @@
+              if (previous == null) {
+                eof = true;
+                return false;
+              }
+              List<Hash> page = nextPage.apply(previous);
+              previous = null;
+              if (!page.isEmpty()) {
+                currentBatch = fetcher.apply(ctx, page).iterator();
+              } else {
+                eof = true;
+                return false;
+              }
+            }
+            T v = currentBatch.next();
+            if (v != null) {
+              consumer.accept(v);
+              previous = v;
+            }
+            return true;
+          }
+        };
+    return StreamSupport.stream(split, false);
+  }
+
+  /**
+   * Builds a {@link CommitLogEntry} using the given values. This function also includes an {@link
+   * EmbeddedKeyList}, if triggered by the values of {@code currentKeyListDistance} and {@link
+   * DatabaseAdapterConfig#getKeyListDistance()}, so read operations may happen.
+   */
+  protected CommitLogEntry buildIndividualCommit(
+      OP_CONTEXT ctx,
+      long timeInMicros,
+      List<Hash> parentHashes,
+      ByteString commitMeta,
+      List<KeyWithBytes> puts,
+      List<Key> unchanged,
+      List<Key> deletes,
+      int currentKeyListDistance) {
+    Hash commitHash = individualCommitHash(parentHashes, commitMeta, puts, unchanged, deletes);
+
+    int keyListDistance = currentKeyListDistance + 1;
+
+    CommitLogEntry entry =
+        CommitLogEntry.of(
+            timeInMicros,
+            commitHash,
+            parentHashes,
+            commitMeta,
+            puts,
+            unchanged,
+            deletes,
+            keyListDistance,
+            null);
+
+    if (keyListDistance >= config.getKeyListDistance()) {
+      entry = buildKeyList(ctx, entry);
+    }
+    return entry;
+  }
+
+  /** Calculate the hash for the contents of a {@link CommitLogEntry}. */
+  @SuppressWarnings("UnstableApiUsage")
+  protected Hash individualCommitHash(
+      List<Hash> parentHashes,
+      ByteString commitMeta,
+      List<KeyWithBytes> puts,
+      List<Key> unchanged,
+      List<Key> deletes) {
+    Hasher hasher = newHasher();
+    hasher.putLong(COMMIT_LOG_HASH_SEED);
+    parentHashes.forEach(h -> hasher.putBytes(h.asBytes().asReadOnlyByteBuffer()));
+    hasher.putBytes(commitMeta.asReadOnlyByteBuffer());
+    puts.forEach(
+        e -> {
+          hashKey(hasher, e.getKey());
+          hasher.putString(e.getContentsId(), StandardCharsets.UTF_8);
+          hasher.putBytes(e.getValue().asReadOnlyByteBuffer());
+        });
+    unchanged.forEach(e -> hashKey(hasher, e));
+    deletes.forEach(e -> hashKey(hasher, e));
+    return Hash.of(UnsafeByteOperations.unsafeWrap(hasher.hash().asBytes()));
+  }
+
+  /** Adds a complete key-list to the given {@link CommitLogEntry}, will read from the database. */
+  protected CommitLogEntry buildKeyList(OP_CONTEXT ctx, CommitLogEntry unwrittenEntry) {
```

Comment:
How is the Key list stored/modelled? I am not 100% clear on that. It was effectively a linked-list in the original impl is it the same? The original impl also ensured that the list was semi-sorted

1. It can't only live on the commit right? 
2. Is it a separate object in the database?
3. Does every op have to read the entire list?

Reply:
The (current) concept to deal with the problem of getting a "full key list quickly" is this:
* The full key list (think: call to `DatabaseAdapter.keys()` / `VersionStore.getKeys`) is constructed from all puts+deletes from all commits
* Since the above is quite expensive, the implementation writes a "full key list" with every N-th commit (`DatabaseAdapterConfig.getKeyListDistance()`, defaults to `20`).

The above (full-key-list within a commit-entry) runs into the same problem as the current TVS implementation (if it becomes too big, the underlying database "runs into trouble"). The plan is to tackle this problem in a follow-up by limiting the number of keys per embedded-key-list and put "overflows" into separate database objects/rows (the commented `EmbeddedKeyList.getKeyListsIds()`). Probably limit the `CommitLogEntry.embeddedKeyList` to maybe a few dozen entries and then overflow to separate objects/rows. Another idea is to manage both the `CommitLogEntry.embeddedKeyList` and its "overflows" in an LRU-ish way - i.e. most recently touched keys end in `CommitLogEntry.embeddedKeyList`, older ones get "spilled out" into the "overflows".

Reply:
I am not sure I want to make the re-implementation less stable than the current one. The current one only runs into trouble if the embedded key list is rather large as it stores pointers to sub-lists of keys. I dont remember the number but its quite a bit larger than teh largest warehouses we would expect, This impl is limited to 10s or 100s of keys before it runs into trouble.

The algorithm you suggest is similar to how the TVS works today (at least in my understanding). The most often changed tables will be near the 'top' of the list of key-lists and the least often changed will sink to the bottom.

Reply:
> This impl is limited to 10s or 100s of keys before it runs into trouble.

It's limited to the max size of a "db object" (-> e.g. the 400k limit in dynamo).

Reply:
agreed. which is orders of magnitude smaller than the current impl and for large table namespaces or delta tables that could be in the 100s of table names